activation/post_supervisor_test.go

```diff
@@ -18,27 +23,86 @@ func Test_PostSupervisor_ErrorOnMissingBinary(t *testing.T) {
 	cmdCfg := DefaultTestPostServiceConfig()
 	cmdCfg.PostServiceCmd = "missing"
 	postCfg := DefaultPostConfig()
-	postOpts := DefaultPostSetupOpts()
 	provingOpts := DefaultPostProvingOpts()
 
-	ps, err := NewPostSupervisor(log.Named("supervisor"), cmdCfg, postCfg, postOpts, provingOpts)
+	ps, err := NewPostSupervisor(log.Named("supervisor"), cmdCfg, postCfg, provingOpts, nil, nil)
 	require.ErrorContains(t, err, "post service binary not found")
 	require.Nil(t, ps)
 }
 
 func Test_PostSupervisor_StopWithoutStart(t *testing.T) {
 	log := zaptest.NewLogger(t)
 
+	cmdCfg := DefaultTestPostServiceConfig()
+	postCfg := DefaultPostConfig()
+	provingOpts := DefaultPostProvingOpts()
+
+	ps, err := NewPostSupervisor(log.Named("supervisor"), cmdCfg, postCfg, provingOpts, nil, nil)
+	require.NoError(t, err)
+	require.NotNil(t, ps)
+
+	require.NoError(t, ps.Stop(false))
+}
+
+func Test_PostSupervisor_Start_FailPrepare(t *testing.T) {
+	log := zaptest.NewLogger(t)
+
 	cmdCfg := DefaultTestPostServiceConfig()
 	postCfg := DefaultPostConfig()
 	postOpts := DefaultPostSetupOpts()
 	provingOpts := DefaultPostProvingOpts()
 
-	ps, err := NewPostSupervisor(log.Named("supervisor"), cmdCfg, postCfg, postOpts, provingOpts)
+	mgr := NewMockpostSetupProvider(gomock.NewController(t))
+	testErr := errors.New("test error")
+	mgr.EXPECT().PrepareInitializer(postOpts).Return(testErr)
+
+	ps, err := NewPostSupervisor(log.Named("supervisor"), cmdCfg, postCfg, provingOpts, mgr, nil)
 	require.NoError(t, err)
 	require.NotNil(t, ps)
 
-	require.NoError(t, ps.Stop())
+	require.ErrorIs(t, ps.Start(postOpts), testErr)
+}
+
+type fatalHook struct {
+	called bool
+}
+
+func (f *fatalHook) OnWrite(*zapcore.CheckedEntry, []zapcore.Field) {
+	f.called = true
+}
+
+func calledFatal(tb testing.TB) zapcore.CheckWriteHook {
+	hook := &fatalHook{}
+	tb.Cleanup(func() { assert.True(tb, hook.called, "zap.Fatal not called") })
+	return hook
+}
+
+func Test_PostSupervisor_Start_FailStartSession(t *testing.T) {
+	log := zaptest.NewLogger(t, zaptest.WrapOptions(zap.WithFatalHook(calledFatal(t))))
+
+	cmdCfg := DefaultTestPostServiceConfig()
+	postCfg := DefaultPostConfig()
+	postOpts := DefaultPostSetupOpts()
+	provingOpts := DefaultPostProvingOpts()
+
+	ctrl := gomock.NewController(t)
+	mgr := NewMockpostSetupProvider(ctrl)
+	mgr.EXPECT().PrepareInitializer(postOpts).Return(nil)
+	mgr.EXPECT().StartSession(gomock.Any()).Return(errors.New("failed start session"))
+
+	sync := NewMocksyncer(ctrl)
+	sync.EXPECT().RegisterForATXSynced().DoAndReturn(func() <-chan struct{} {
+		ch := make(chan struct{})
+		close(ch)
+		return ch
+	})
```

Comment:
This pattern repeats over and over. How about a little helper to increase readability:
```go
func closedChannel() <-chan struct{} {
	ch := make(chan struct{})
	close(ch)
	return ch
}
```
```suggestion
	sync.EXPECT().RegisterForATXSynced().Return(closedChannel())
```

Reply:
I like to repeat myself in tests over factoring out code into helper/utility functions. Test code in my opinion should be dumb and obvious why it fails when it fails. 

Consider a future change where some tests break with `closedChannel()` and some don't. Often this leads to `closedChannel` becoming more complex instead of a second function being added for the changed behaviour in the failing tests. Over multiple iterations this leads to test code that is a) hard to change/extend without breaking existing tests and b) increased complexity where it's not obvious any more what is happening in a test.

This here is trivial enough, that I don't mind adding a helper function if you insist, but I'd prefer not to 🙂

Reply:
In general, I agree with the statement that test helpers often become too big and do too much, which results in difficult future refactoring or adding new tests. But on the other hand, this helper would be super simple. What could be broken there realistically? The benefit is that the test code is not bloated with unnecessary code. `closedChannel()` represents the intention nicely.